blockchain/internal/spendpruner/pruner.go

```diff
@@ -0,0 +1,292 @@
+	// block hash.
+	err := s.db.Update(func(tx database.Tx) error {
+		err := dbUpdateSpendConsumerDeps(tx, *blockHash, dependents)
+		return err
+	})
+	if err != nil {
+		return fmt.Errorf("unable to update persisted consumer "+
+			"dependencies for block hash %v: %v", blockHash, err)
+	}
+
+	return nil
+}
+
+// RemoveSpendConsumerDependency removes the provided spend consumer dependency
+// associated with the provided block hash from the spend pruner. The block
+// hash is removed as a key of the dependents map once all its dependency
+// entries are removed.
+func (s *SpendJournalPruner) RemoveSpendConsumerDependency(blockHash *chainhash.Hash, consumerID string) {
+	s.dependentsMtx.Lock()
+	defer s.dependentsMtx.Unlock()
+
+	dependents, ok := s.dependents[*blockHash]
+	if !ok {
+		// No entry for block hash found, do nothing.
+		return
+	}
+
+	for idx := 0; idx < len(dependents); idx++ {
+		if dependents[idx] == consumerID {
+			dependents = append(dependents[:idx], dependents[idx+1:]...)
+			s.dependents[*blockHash] = dependents
+			break
+		}
+	}
+
+	if len(dependents) == 0 {
+		delete(s.dependents, *blockHash)
+		s.pruneCh <- blockHash
+	}
+}
+
+// removeSpendEntry removes the associated spend data of the provided block
+// hash from the prune set as well as removing all consumer dependencies
+// associated.
+func (s *SpendJournalPruner) removeSpendEntry(blockHash *chainhash.Hash) {
+	s.dependentsMtx.Lock()
+	defer s.dependentsMtx.Unlock()
+
+	delete(s.dependents, *blockHash)
+
+	// Remove the tracked spend journal entry for the provided
+	// block hash.
+	err := s.db.Update(func(tx database.Tx) error {
+		err := dbUpdateSpendConsumerDeps(tx, *blockHash, nil)
+		return err
+	})
+	if err != nil {
+		log.Errorf("unable to remove persisted consumer dependencies "+
+			"entry for block hash %v: %v", blockHash, err)
+	}
+}
+
+// loadSpendConsumerDeps loads persisted consumer dependencies from the database.
+func (s *SpendJournalPruner) loadSpendConsumerDeps() error {
+	return s.db.View(func(tx database.Tx) error {
+		consumerDeps, err := dbFetchSpendConsumerDeps(tx)
+		if err != nil {
+			return err
+		}
+
+		s.dependentsMtx.Lock()
+		for k, v := range consumerDeps {
+			s.dependents[k] = v
+		}
+		s.dependentsMtx.Unlock()
+
+		return nil
+	})
+}
+
+// handleSignals processes incoming signals to the spend pruner.
+func (s *SpendJournalPruner) handleSignals(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			s.wg.Done()
+			return
+
+		case blockHash := <-s.pruneCh:
+			err := s.chain.RemoveSpendEntry(blockHash)
+			if err != nil {
+				log.Errorf("unable to prune spend data for "+
+					"block hash (%s): %v", blockHash, err)
+			}
+
+		case blockHash := <-s.connectedCh:
+			if !s.DependencyExists(blockHash) {
+				continue
+			}
+
+			// Remove the prune entry of the provided connected block hash
+			// and its consumer dependencies.
+			s.removeSpendEntry(blockHash)
```

Comment:
If I'm reading this correctly, this is removing a spend journal entry when a block is connected?  The spend journal entry should only be removed after `BlockChain` _disconnects_ a block (and consumers no longer need it).

Reply:
Yeah that was a bad function name to use there.  It's actually only removing the key/value pair of the spend consumer dependencies and the associated connected block from the prune set since block connection invalidates the need to prune the spend journal entry of the block. I've renamed the function to `removeSpendConsumerDeps`, good catch.

Reply:
Got it, thanks for clarifying.  Renaming it definitely helps.  Planning on reviewing the latest changes more in depth soon yet just noticed this when taking a quick look.